src/doc/if-let-guard-review.md

```diff
@@ -0,0 +1,268 @@
+succeed, the body can use both `x` and `y`.  If either fails (pattern fails or guard fails), this arm is skipped.
+
+One important restriction is that a single match arm cannot have two `if`-guards.  That is, you cannot write something like 
+`PAT if cond1 if let P = E => ...` with two separate `if`s.  You may combine a normal boolean condition with a `let` 
+by chaining with `&&`, but only one `if` keyword is allowed.  The RFC explicitly states “An arm may not have both an 
+`if` and an `if let` guard” (i.e. you can’t do `if cond && let ...` *and* then another `if`, etc.).  
+(You *can* do something like `if let P = E && cond` by writing `if let P = E && cond =>`, treating the boolean as part 
+of a let-chain, but that is a single `if` in syntax.)
+
+## Feature Gate and Errors
+
+As of now, `if let` guards are still unstable. The compiler requires the feature flag `#![feature(if_let_guard)]` to enable them. 
+If one uses an `if let` guard without the feature, one gets an error similar to:
+
+```
+error[E0658]: `if let` guards are experimental
+   |
+LL |     _ if let true = true => {}
+   |        ^^^^^^^^^^^^^^^^
+   = help: you can write `if matches!(<expr>, <pattern>)` instead of `if let <pattern> = <expr>`
+```
+
+This message is verified by the compiler’s test suite (e.g. `feature-gate.rs`) and comes from the feature-gate code in the parser.  
+The tests also ensure the old (`let`-in-`if` without the feature) error is preserved. For example:
+
+```rust
+match () {
+    () if true && let 0 = 1 => {}      // error: `let` expressions are unstable (since no feature)
+    () if let 0 = 1 && true => {}      // error: `if let` guards are experimental
+    _ => {}
+}
+```
+
+The test suite checks that these errors mention both the unstable-let and the experimental guard exactly as above.  
+Once the feature is stabilized, these errors will no longer appear.
+
+## Temporaries and Drop Order
+
+A subtle aspect of `if let` guards is the handling of temporaries (and destructor calls) within the guard expression. 
+The Rust reference explains that a *match guard* creates its own temporary scope: any temporaries produced by `GUARD_EXPR` 
+live only until the guard finishes evaluating. Concretely, this means:
+
+* The `guard_expr` is evaluated *after* matching `PAT` but *before* executing the arm’s body (if taken).
+* Any temporary values created during `guard_expr` are dropped immediately after the guard’s scope ends (i.e. before entering the arm body).
+* If the guard fails, those temporaries are dropped right then, and the compiler proceeds to the next arm.
+
+In effect, the drop semantics are the same as for an ordinary match guard or an `if let` in an `if` expression: no unexpected 
+extension of lifetimes. (In Rust 2024 edition, there is a finer rule that even in `if let` 
+expressions temporaries drop before the `else` block; but for match guards the effect is that temporaries from the 
+guard are dropped before the arm body.)
```

Comment:
I think we should take the approach that we adopt 2024 drop order behaviour on all editions. As it's a green field new feature, we can do that. Because I see that there is edition differences on the example from https://github.com/rust-lang/rust/issues/104843 adapted to if let guards: there is a broken MIR ICE on 2021, but no ICE on 2024.

When it comes to https://github.com/rust-lang/rust/issues/103476#issuecomment-1298155838 , I can't reproduce it on the 2021 edition.